src/wallet/rpcwallet.cpp

```diff
@@ -161,8 +161,8 @@ UniValue getnewaddress(const JSONRPCRequest& request)
 
     // Generate a new key that is added to wallet
     CPubKey newKey;
-    if (!pwallet->GetKeyFromPool(newKey)) {
-        throw JSONRPCError(RPC_WALLET_KEYPOOL_RAN_OUT, "Error: Keypool ran out, please call keypoolrefill first");
+    if (!pwallet->GetKeyFromPool(newKey) || (pwallet->IsHDEnabled() && !pwallet->HasUnusedKeys(GetArg("-keypoolcritical", DEFAULT_KEYPOOL_CRITICAL)))) {
```

Comment:
I originally wasn't doing this because `GetKeyFromPool()` calls into `ReserveKeyFromKeyPool()` which tops up the keypool. I didn't want to fail `GetKeyFromPool()` when the wallet is unlocked and able to top up the keypool.

I've now added a `TopUpKeyPool()` call in `GetKeyFromPool()`, so we try to top up before checking that the keypool hasn't fallen to the critical level.